subprojects/core/src/main/java/org/gradle/internal/buildtree/DefaultBuildTreeModelCreator.java

```diff
@@ -78,5 +104,71 @@ public boolean queryModelActionsRunInParallel() {
         public <T> List<T> runQueryModelActions(List<Supplier<T>> actions) {
             return actionRunner.run(actions);
         }
+
+        private ToolingModelScope locateBuilderForDefaultTarget(String modelName, boolean param) {
+            return locateBuilderForTarget(defaultTarget, modelName, param);
+        }
+
+        private ToolingModelScope locateBuilderForTarget(BuildState target, String modelName, boolean param) {
+            return target.withToolingModels(controller -> controller.locateBuilderForTarget(modelName, param));
+        }
+
+        private ToolingModelScope locateBuilderForTarget(ProjectState target, String modelName, boolean param) throws UnknownModelException {
+            return target.getOwner().withToolingModels(controller -> controller.locateBuilderForTarget(target, modelName, param));
+        }
+
+        @Nullable
+        private Object getModelForScope(ToolingModelScope scope, String modelName, @Nullable Object parameter) {
+            if (parameter == null) {
+                return scope.getModel(modelName, null);
+            } else {
+                return scope.getModel(modelName, parameterCarrierFactory.createCarrier(parameter));
+            }
+        }
+
+        private String describeTarget(@Nullable BuildTreeModelTarget target) {
+            if (target == null) {
+                return "default scope";
+            } else if (target.getProjectPath() != null) {
+                return "project scope";
+            } else {
+                return "build scope";
+            }
+        }
+
+        private ToolingModelScope getTarget(@Nullable BuildTreeModelTarget target, String modelName, boolean parameter) {
+            if (target == null) {
+                return locateBuilderForDefaultTarget(modelName, parameter);
+            }
+
+            BuildState build = findBuild(target);
+            Path projectPath = target.getProjectPath();
+            if (projectPath == null) {
+                return locateBuilderForTarget(build, modelName, parameter);
+            } else {
+                ProjectState project = findProject(build, projectPath);
+                return locateBuilderForTarget(project, modelName, parameter);
+            }
+        }
+
+        private BuildState findBuild(BuildTreeModelTarget target) {
+            File targetBuildRootDir = target.getBuildRootDir();
+            AtomicReference<BuildState> match = new AtomicReference<>();
+            buildStateRegistry.visitBuilds(buildState -> {
+                if (buildState.isImportableBuild() && buildState.getBuildRootDir().equals(targetBuildRootDir)) {
+                    match.set(buildState);
```

Comment:
🚀 Any chance of breaking the loop here?

Reply:
The `visitBuilds` method does not support stopping early. Because it has many call-sites, I wouldn't try to change it in the same PR. For the record, this code is not new and has been moved here from `DefaultBuildController`